### PR TITLE
Use filter metadata.

### PR DESCRIPTION
**Pull request by notbdu**
Leverage filter metadata in queries.

### CODE REVIEW

File: values/decoding/fs_bool_values.go [hunk 1]
```diff
@@ -38,13 +39,30 @@ func (v *fsBasedBoolValues) Iter() (iterator.ForwardBoolIterator, error) {
...
 
-// TODO(xichen): Filter implementation should take advantage of the metadata
-// to do more intelligent filtering, e.g., checking if the value is within the
-// value range.
 func (v *fsBasedBoolValues) Filter(
 	op filter.Op,
 	filterValue *field.ValueUnion,
 ) (iterator.PositionIterator, error) {
+	var (
+		numTrue  = v.Metadata().NumTrues
```
Comment (xichen2020) [inline values/decoding/fs_bool_values.go]: `numTrues`, also you can just use `v.metaProto.NumTrues` to save the cost of a function call.

File: values/decoding/fs_bool_values.go [hunk 2]
```diff
@@ -38,13 +39,30 @@ func (v *fsBasedBoolValues) Iter() (iterator.ForwardBoolIterator, error) {
...
+		numFalse = v.Metadata().NumFalses
```
Comment (xichen2020) [inline values/decoding/fs_bool_values.go]: `numFalses`, and same comment as above.

File: values/decoding/fs_bool_values.go [hunk 3]
```diff
@@ -38,13 +39,30 @@ func (v *fsBasedBoolValues) Iter() (iterator.ForwardBoolIterator, error) {
...
+	)
+	switch op {
```
Comment (xichen2020) [inline values/decoding/fs_bool_values.go]: Probably safer to do a nil check `filterValue` as well as checking `filterValue.Type == field.BoolType` first before the `switch`, and return an error early otherwise.

File: values/decoding/fs_string_values.go [hunk 4]
```diff
@@ -60,14 +61,38 @@ func (v *fsBasedStringValues) Iter() (iterator.ForwardStringIterator, error) {
 	return newStringIteratorFromMeta(v.metaProto, v.encodedValues, v.dictArr, v.encodedDictBytes)
 }
 
-// TODO(xichen): Filter implementation should take advantage of the metadata
-// to do more intelligent filtering, e.g., checking if the value is within the
-// value range, intelligently look up filter values and bail early if not found,
-// perform filtering directly against the index to avoid string comparisons.
+// TODO(xichen): Filter implementation should intelligently look up filter values and bail early if not found
```
Comment (xichen2020) [inline values/decoding/fs_string_values.go]: Perhaps handle the rest of the TODOs as well?

Comment (notbdu) [reply values/decoding/fs_string_values.go]: The rest of the TODOs are more involved. Let's save them for a future PR and get these more minor changes in now.

**Comment by black-adder** [comment]
@xichen2020 updated PR; that's an impressive decrease in coverage... I'll add some tests

File: values/decoding/fs_string_values.go [hunk 5]
```diff
@@ -60,14 +61,35 @@ func (v *fsBasedStringValues) Iter() (iterator.ForwardStringIterator, error) {
...
+	}
+	if !op.StringIsInRange(v.metaProto, filterValue.StringVal) {
+		return impl.NewEmptyPositionIterator(), nil
+	}
+	if v.metaProto.Encoding == encodingpb.EncodingType_DICTIONARY {
+		idx, ok := v.dictMap[filterValue.StringVal]
+		if !ok {
+			return impl.NewEmptyPositionIterator(), nil
+		}
+		idxIterator, err := newIndexIteratorFromMeta(v.metaProto, v.encodedValues, v.encodedDictBytes)
```
Comment (xichen2020) [inline values/decoding/fs_string_values.go]: Noice

File: values/decoding/fs_time_values.go [hunk 6]
```diff
@@ -36,19 +38,41 @@ func (v *fsBasedTimeValues) Metadata() values.TimeValuesMetadata {
...
+	if filterValue == nil {
+		return nil, errNilFilterValue
+	}
+	if filterValue.Type != field.TimeType {
+		return nil, errUnexpectedFilterValueType
+	}
+	if !op.TimeIsInRange(v.metaProto, filterValue.TimeNanosVal) {
+		return impl.NewEmptyPositionIterator(), nil
+	}
+	// TODO(wjang): newTimeIteratorFromMeta below also calls protoResolutionToDuration. The overhead shouldn't
```
Comment (xichen2020) [inline values/decoding/fs_time_values.go]: `TODO:(wjang)` -> `NB(wjang)` unless you plan to fix this later.

Alternatively you can have a `newTimeIteratorWithResolution` that takes in a resolution and a meta.

File: values/decoding/time_decode.go [hunk 7]
```diff
@@ -45,6 +45,7 @@ func (dec *timeDecoder) DecodeRaw(data []byte) (values.CloseableTimeValues, erro
 func newTimeIteratorFromMeta(
 	metaProto encodingpb.TimeMeta,
 	encodedBytes []byte,
+	scaleUp bool,
```
Comment (xichen2020) [inline values/decoding/time_decode.go]: Make this an enum perhaps? I'm a fan of more explicit parameters, e.g.,
```
type scaleMode int

const (
  noScale scaleMode = iota
  withScale
)
```
Then you can call this with `newTimeIterator(..., noScale)` which is more clear than `true` or `false`.

File: values/decoding/fs_time_values.go [hunk 8]
```diff
@@ -36,19 +38,41 @@ func (v *fsBasedTimeValues) Metadata() values.TimeValuesMetadata {
...
+	// be too bad compared to the time spent iterating.
+	resolution, err := protoResolutionToDuration(v.metaProto.Resolution)
+	if err != nil {
+		return nil, err
+	}
+	timeIterator, err := newTimeIteratorFromMeta(v.metaProto, v.encodedValues, false)
+	if err != nil {
+		return nil, err
+	}
+	scaledDownFilterValue := field.NewTimeUnion(convert.ScaleDownTimeFn(filterValue.TimeNanosVal, resolution))
```
Comment (xichen2020) [inline values/decoding/fs_time_values.go]: Add some comments here perhaps?

File: values/encoding/int_encode.go [hunk 9]
```diff
@@ -96,7 +96,7 @@ func (enc *intEncoder) Encode(
 	}
 
 	if len(dictionary) <= maxCardinalityAllowed {
-		bytesPerDictionaryValue := int64(math.Ceil(float64(bits.Len(uint(valueMeta.Max-valueMeta.Min))) / 8.0))
+		bytesPerDictionaryValue := int64(math.Max(math.Ceil(float64(bits.Len(uint(valueMeta.Max-valueMeta.Min)))/8.0), 1.0))
```
Comment (black-adder) [inline values/encoding/int_encode.go]: without this, if the range is 0, `bytesPerDictValue` is 0 and dictionary decoding will fail. Somehow all the dictionary encoding tests passed due to some miracle.

Also, could we encode this with bits instead of bytes?

File: values/decoding/int_decode.go [hunk 10]
```diff
@@ -43,8 +44,20 @@ func (dec *intDecoder) DecodeRaw(data []byte) (values.CloseableIntValues, error)
 	if err != nil {
 		return nil, err
 	}
-
-	return newFsBasedIntValues(metaProto, data[bytesRead:], dec.dictionaryProto.Data, nRead), nil
+	var dict []int
+	if nRead > 0 {
```
Comment (xichen2020) [inline values/decoding/int_decode.go]: Move this logic to `tryDecodeIntDictionary` and return `[]int` from there?

File: values/encoding/int_encode.go (hunk 9 above)
Comment (xichen2020) [inline values/encoding/int_encode.go]: Would you mind adding a few tests to make sure encoding/decoding/roundtrip an empty int/string value set works as expected? Thanks!

File: values/decoding/int_decode.go [hunk 11]
```diff
@@ -65,21 +52,34 @@ func tryDecodeIntDictionary(
...
+		return 0, nil, nil
 	case encodingpb.EncodingType_DICTIONARY:
 		reader := bytes.NewReader(data)
 		bytesRead, err := xproto.DecodeIntDictionary(reader, dictionaryProto, decodeBuf)
 		if err != nil {
-			return 0, err
+			return 0, nil, err
+		}
+		var dict []int
+		if bytesRead > 0 {
```
Comment (xichen2020) [inline values/decoding/int_decode.go]: nit: reduce nesting w/
```
if bytesRead == 0 {
  return 0, nil, nil
}
...
```

File: values/decoding/int_decode.go [hunk 12]
```diff
@@ -112,3 +112,20 @@ func newIntIteratorFromMeta(
 		return nil, fmt.Errorf("invalid int encoding type: %v", metaProto.Encoding)
 	}
 }
+
+// newDictionaryIndexIterator creates a new int iterator that can be used to iterate over
+// dictionary index values.
+// NB: EncodingType must be encodingpb.EncodingType_DICTIONARY when calling this function.
+func newDictionaryIndexIterator(
```
Comment (xichen2020) [inline values/decoding/int_decode.go]: `newIntDictionaryIndexIterator` perhaps?

File: values/decoding/string_decode.go [hunk 13]
```diff
@@ -113,10 +113,10 @@ func newStringIteratorFromMeta(
 	}
 }
 
-// newIndexIteratorFromMeta creates a new int iterator from string metadata that can be used
+// newDictionaryIndexIteratorFromMeta creates a new int iterator from string metadata that can be used
 // to iterate over dictionary index values.
 // NB: EncodingType must be encodingpb.EncodingType_DICTIONARY when calling this function.
-func newIndexIteratorFromMeta(
+func newDictionaryIndexIteratorFromMeta(
```
Comment (xichen2020) [inline values/decoding/string_decode.go]: `newStringDictionaryIndexIterator` perhaps?
